### DISCUSSION TITLE
Adding Custom, time-dependent MACH_NUMBER

### DISCUSSION THREAD

**Discussion by thunderqd** [su2code/SU2#1836] [Q&A]
Hi all,

I'm new to CFD and SU2. I'm trying to modify the c++ code to add a time-dependent mach in the compressible Euler equations (SU2 Solver: EULER) for a rocket head cone design. At about line 786 of the CEulerSolver.cpp, I want to have a slight numerical change in its mach, for example, a sine function.

After I saw the example of the Adding Custom Source Terms to Euler Equations I modified the code like this:

```
  const auto dt           = config->GetDelta_UnstTimeND();
  su2double time          = config->GetTime_Step() * dt;
  su2double Alpha         = config->GetAoA()*PI_NUMBER/180.0;
  su2double Beta          = config->GetAoS()*PI_NUMBER/180.0;
  const su2double MachValue     = config->GetMach();
  su2double Mach          = 2 * MachValue + 0.01 * sin(2 * PI_NUMBER * time);
```

But during postprocessing I noticed that Mach becomes 2 * MachValue + 0 as the "time" could not be updated to the series of 0, 1, 2, 3, 4... So I'm wondering how I can build a mach that could change over time. I'm not sure what kind of syntax I should be using or how exactly to implement this mach term. 

I am using SU2 7.3.1. Thank you so much!

**Comment by bigfooted** [comment]
I think you need config->GetTimeIter() to get the iteration number, and multiply that by the timestep dt

**Reply by thunderqd** [reply]
Many thanks for your answer.

I am really sorry I already tried this method, according to the example in the following link:

[](https://www.cfd-online.com/Forums/su2/227968-adding-custom-source-terms-euler-equations.html)

`su2double time = config->GetTimeIter() * dt;`

But the value of "time" was still 0 in any time step, ie

`0.01 * sin(2 * PI_NUMBER * 0);`

I'm a bit unsure if it's an issue about the declaration of the constant/variable.

I would be really grateful if there is a current time expression method for a serie such as 0, 1, 2, 3, 4....

**Reply by thunderqd** [reply]
Is there any way to solve this problem? The timestep is always 0 and cannot be updated. In addition to Mach, it is also possible to modify the inlet boundary velocity to custom speed.
If anyone has encountered a similar problem, I would be very thanksful for your advice.

**Reply by thunderqd** [reply]
Even the constant value dt is zero too (which is GetDelta_UnstTimeND())
if I trying to type:
`su2double Mach = GetMach() * dt * 100;`
the mach output is still zero
I would like to know if it was declared as 0 at the beginning, which caused the time not to be updated?
`Delta_UnstTimeND = 0.0`

**Reply by bigfooted** [reply]
Do you want to update a boundary condition? If so, then you should do so in the respective function for the boundary condition, e.g. BC_Far_Field if it is a far-field boundary, or BC_Inlet if it is an inlet.

**Reply by thunderqd** [reply]
The primary goal of my simulation is to define a freestream velocity that could vary slightly (vibrate like a wave) over time. Regarding this 2D airfoil simulation, I remember that some studies have used the inlet boundary condition to define the air velocity (but no design optimization by fluent), but SU2 cannot import udf, so I am trying to hardcode.

However, the main problem is no matter which function is defined, the current simulation time cannot be updated (mainly because the declared new variables cannot be recognized), and even newly defined constants such as dt cannot be recognized, so these values are 0 in any timestep.

Today I will try to define air velocity with BC_Inlet. If the updated BC_Inlet of defining variable air flow rate I can use it.

**Reply by bigfooted** [reply]
If I print TimeIter() inside BC_Far_Field, I get the current iteration. You can then modify the boundary conditions through the free-stream velocity. 

You can create user defined functions using the python wrapper, once you have this working 'hard-coded' it would be good to generalize by adding this functionality to the python wrapper.
